jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp

```diff
@@ -0,0 +1,164 @@
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/o2r other materials provided
+ *     with the distribution.
+ *   * Neither the name of the JSK Lab nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+/*
+ * container_occupancy_detector_nodelet.cpp
+ * Author: Yoshiki Obinata <obinata@jsk.imi.i.u-tokyo.ac.jp>
+ */
+
+#include "jsk_pcl_ros/container_occupancy_detector.h"
+
+namespace jsk_pcl_ros{
+
+    void ContainerOccupancyDetector::onInit(){
+        jsk_topic_tools::ConnectionBasedNodelet::onInit();
+        boxes_occupancy_pub_
+            = advertise<jsk_recognition_msgs::BoundingBoxArray>(*pnh_, "container/occupancies", 1);
+        tf_listener_ = new tf2_ros::TransformListener(tf_buffer_);
+        onInitPostProcess();
+    }
+
+    ContainerOccupancyDetector::~ContainerOccupancyDetector(){
+        // message_filters::Synchronizer needs to be called reset
+        // before message_filters::Subscriber is freed.
+        // Calling reset fixes the following error on shutdown of the nodelet:
+        // terminate called after throwing an instance of
+        // 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::lock_error> >'
+        //     what():  boost: mutex lock failed in pthread_mutex_lock: Invalid argument
+        // Also see https://github.com/ros/ros_comm/issues/720 .
+        sync_.reset();
+    }
+
+    void ContainerOccupancyDetector::subscribe(){
+        sub_boxes_.subscribe(*pnh_, "container/boxes", 1);
+        sub_points_.subscribe(*pnh_, "container/points", 1);
+        sub_point_indices_.subscribe(*pnh_, "container/point_indices", 1);
+        sync_ = std::make_shared<message_filters::Synchronizer<SyncPolicy> >(100);
```

Comment:
Adding an approximate sync option is a good direction for other use cases.

Reply:
Thanks, I added